File: imagelab-frontend/src/types/pipeline.ts
export interface PipelineStep {
  block_id?: string;
  type: string;
  params: Record<string, unknown>;
}

export interface PipelineRequest {
  image: string;
  image_format: string;
  pipeline: PipelineStep[];
}

export interface StepTiming {
  step: number;
  operator_type: string;
  duration_ms: number;
}

export interface PipelineTimings {
  total_ms: number;
  steps: StepTiming[];
}

export interface StepResult {
  index: number;
  block_id?: string | null;
  type: string;
  success: boolean;
  thumbnail?: string | null;
  image_format?: string | null;
  timing_ms?: number | null;
  has_full_image: boolean;
  error?: string | null;
}

export interface ImageAnalysis {
  width: number;
  height: number;
  channels: number;
  dtype: string;
  min: number;
  max: number;
  mean: number | number[];
  std: number | number[];
}

export interface PipelineResponse {
  success: boolean;
  execution_id?: string | null;
  image?: string;
  image_format?: string;
  error?: string;
  step?: number;
  error_block_id?: string | null;
  timings?: PipelineTimings;
  step_results?: StepResult[];
}

export interface StepInspectResponse {
  success: boolean;
  execution_id: string;
  block_id: string;
  index: number;
  type: string;
  image: string;
  image_format: string;
  timing_ms?: number | null;
  analysis: ImageAnalysis;
}

File: imagelab-backend/app/models/pipeline.py
from pydantic import BaseModel, Field, model_validator

from app.models.graph import PipelineGraph


class PipelineStep(BaseModel):
    type: str
    block_id: str | None = None
    params: dict = Field(default_factory=dict)
    branches: dict[str, list["PipelineStep"]] = Field(default_factory=dict)


class PipelineRequest(BaseModel):
    image: str
    image_format: str = "png"
    graph: PipelineGraph | None = None
    # Keep default None so Pydantic catches when the key is completely missing in raw JSON
    pipeline: list[PipelineStep] | None = None

    @model_validator(mode="after")
    def validate_payload_presence(self):
        # 1. If 'pipeline' key was completely missing and no 'graph' was given, raise ValueError (HTTP 422)
        if self.pipeline is None and self.graph is None:
            raise ValueError("Field 'pipeline' or 'graph' is required.")
        # 2. If 'pipeline' key was explicitly provided as missing/None, initialize to [] for execution
        if self.pipeline is None:
            self.pipeline = []
        return self


class StepTiming(BaseModel):
    step: int
    operator_type: str
    duration_ms: float


class PipelineTimings(BaseModel):
    total_ms: float
    steps: list[StepTiming]


class StepResult(BaseModel):
    index: int
    block_id: str | None = None
    type: str
    success: bool
    thumbnail: str | None = None
    image_format: str | None = None
    timing_ms: float | None = None
    has_full_image: bool = False
    error: str | None = None


class ImageAnalysis(BaseModel):
    width: int
    height: int
    channels: int
    dtype: str
    min: float
    max: float
    mean: float | list[float]
    std: float | list[float]


class ImageHistogram(BaseModel):
    bins: list[int]
    luminance: list[int]
    red: list[int] | None = None
    green: list[int] | None = None
    blue: list[int] | None = None


class PipelineResponse(BaseModel):
    success: bool
    execution_id: str | None = None
    image: str | None = None
    image_format: str | None = None
    error: str | None = None
    step: int | None = None
    error_block_id: str | None = None
    timings: PipelineTimings | None = None
    step_results: list[StepResult] = []


class StepInspectResponse(BaseModel):
    success: bool
    execution_id: str
    block_id: str
    index: int
    type: str
    image: str
    image_format: str
    timing_ms: float | None = None
    analysis: ImageAnalysis
    histogram: ImageHistogram
